quisp/modules/QRSA/RuleEngine/RuleEngine_test.cc

```diff
@@ -16,53 +18,104 @@ using namespace omnetpp;
+};
+
 class Strategy : public quisp_test::TestComponentProviderStrategy {
  public:
-  Strategy() : mockQubit(nullptr) {}
-  Strategy(MockStationaryQubit* _qubit) : mockQubit(_qubit) {}
-  ~Strategy() { delete mockQubit; }
+  Strategy() : mockQubit(nullptr), routingDaemon(nullptr), hardwareMonitor(nullptr) {}
+  Strategy(MockStationaryQubit* _qubit, MockRoutingDaemon* routing_daemon, MockHardwareMonitor* hardware_monitor)
+      : mockQubit(_qubit), routingDaemon(routing_daemon), hardwareMonitor(hardware_monitor) {}
+  ~Strategy() {
+    delete mockQubit;
+    delete routingDaemon;
+    delete hardwareMonitor;
+  }
   MockStationaryQubit* mockQubit = nullptr;
+  MockRoutingDaemon* routingDaemon = nullptr;
+  MockHardwareMonitor* hardwareMonitor = nullptr;
   StationaryQubit* getStationaryQubit(int qnic_index, int qubit_index, QNIC_type qnic_type) override {
     if (mockQubit == nullptr) mockQubit = new MockStationaryQubit();
     return mockQubit;
   };
+  IRoutingDaemon* getRoutingDaemon() override { return routingDaemon; };
+  IHardwareMonitor* getHardwareMonitor() override { return hardwareMonitor; };
 };
 
 class RuleEngineTestTarget : public quisp::modules::RuleEngine {
  public:
   using quisp::modules::RuleEngine::initialize;
   using quisp::modules::RuleEngine::par;
-  RuleEngineTestTarget(MockStationaryQubit* mockQubit) : quisp::modules::RuleEngine() {
+  RuleEngineTestTarget(MockStationaryQubit* mockQubit, MockRoutingDaemon* routingdaemon, MockHardwareMonitor* hardware_monitor) : quisp::modules::RuleEngine() {
     setParInt(this, "address", 123);
     setParInt(this, "number_of_qnics_rp", 0);
-    setParInt(this, "number_of_qnics_r", 0);
-    setParInt(this, "number_of_qnics", 0);
-    setParInt(this, "total_number_of_qnics", 0);
-    // setParBool(this, "link_tomography", false);
-    // setParStr(this, "tomography_output_filename", "test_file");
-    // setParStr(this, "file_dir_name", "out/tests");
-    // setParInt(this, "initial_purification", 0);
-    // setParBool(this, "X_purification", true);
-    // setParBool(this, "Z_purification", true);
-    // setParInt(this, "Purification_type", 0);
-    // setParInt(this, "num_measure", 0);
-
+    setParInt(this, "number_of_qnics_r", 1);
+    setParInt(this, "number_of_qnics", 1);
+    setParInt(this, "total_number_of_qnics", 2);
     this->setName("rule_engine_test_target");
-    this->provider.setStrategy(std::make_unique<Strategy>(mockQubit));
+    this->provider.setStrategy(std::make_unique<Strategy>(mockQubit, routingdaemon, hardware_monitor));
   }
+
+ protected:
+  // setter function for allResorces[qnic_type][qnic_index]
+  void setAllResources(int qnic_type, int qnic_index, int partner, StationaryQubit* qubit) { this->allResources[qnic_type][qnic_index].insert(std::make_pair(partner, qubit)); };
+
+ private:
+  FRIEND_TEST(RuleEngineTest, ESResourceUpdate);
+  friend class MockRoutingDaemon;
+  friend class MockHardwareMonitor;
```

Comment:
you don't need to set friend class and FRIEND_TEST. this class should be like this
```cpp
class RuleEngineTestTarget : public quisp::modules::RuleEngine {
 public:
  using quisp::modules::RuleEngine::initialize;
  using quisp::modules::RuleEngine::par;
  using quisp::modules::RuleEngine::updateResources_EntanglementSwapping;

  RuleEngineTestTarget(MockStationaryQubit* mockQubit, MockRoutingDaemon* routingdaemon, MockHardwareMonitor* hardware_monitor) : quisp::modules::RuleEngine() {
    setParInt(this, "address", 123);
    setParInt(this, "number_of_qnics_rp", 0);
    setParInt(this, "number_of_qnics_r", 1);
    setParInt(this, "number_of_qnics", 1);
    setParInt(this, "total_number_of_qnics", 2);
    this->setName("rule_engine_test_target");
    this->provider.setStrategy(std::make_unique<Strategy>(mockQubit, routingdaemon, hardware_monitor));
  }

  // setter function for allResorces[qnic_type][qnic_index]
  void setAllResources(int qnic_type, int qnic_index, int partner, StationaryQubit* qubit) { this->allResources[qnic_type][qnic_index].insert(std::make_pair(partner, qubit)); };
};
```